Compiling for Android failed

Compiling for Android is failing with following output:

```
$ make
In file included from /home/kartz/qpdf/libqpdf/QUtil.cc:16:
/home/kartz/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/fstream:952:20: error: use of undeclared identifier 'ftello'; did you mean 'ftell'?
    pos_type __r = ftello(__file_);
                   ^
/home/kartz/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/c++/v1/cstdio:141:9: note: 'ftell' declared here
using ::ftell;
        ^
/home/kartz/qpdf/libqpdf/QUtil.cc:571:12: error: use of undeclared identifier 'fseeko'; did you mean 'fseek'?
    return fseeko(
           ^~~~~~
           fseek
/home/kartz/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/stdio.h:223:5: note: 'fseek' declared here
int fseek(FILE* __fp, long __offset, int __whence);
    ^
/home/kartz/qpdf/libqpdf/QUtil.cc:573:9: warning: implicit conversion loses integer precision: 'long long' to 'long' [-Wshorten-64-to-32]
        QIntC::IntConverter<qpdf_offset_t, off_t>::convert(offset),
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kartz/qpdf/libqpdf/QUtil.cc:590:29: error: use of undeclared identifier 'ftello'; did you mean 'ftell'?
    return QIntC::to_offset(ftello(stream));
                            ^~~~~~
                            ftell
/home/kartz/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/stdio.h:224:6: note: 'ftell' declared here
long ftell(FILE* __fp);
     ^
1 warning and 3 errors generated.
```

Here is the complete output log of `cmake ./`:

```
$ cmake -DCMAKE_TOOLCHAIN_FILE="${ANDROID_HOME}/build/cmake/android.toolchain.cmake" \
    -DANDROID_ABI=armeabi-v7a \
    -DANDROID_NATIVE_API_LEVEL=21 \
    -DUSE_IMPLICIT_CRYPTO=OFF \
    -DALLOW_CRYPTO_NATIVE=ON \
    -DREQUIRE_CRYPTO_NATIVE=ON \
    -DUSE_INSECURE_RANDOM=ON \
    -DSKIP_OS_SECURE_RANDOM=ON \
    ..
-- The C compiler identification is Clang 14.0.7
-- The CXX compiler identification is Clang 14.0.7
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/kartz/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /home/kartz/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test ATOMIC_WORKED0
-- Performing Test ATOMIC_WORKED0 - Success
-- Performing Test flag_wall
-- Performing Test flag_wall - Success
-- Using -Wall: YES
-- Performing Test flag_conversion
-- Performing Test flag_conversion - Success
-- Using -Wconversion: YES
-- Performing Test flag_sign-conversion
-- Performing Test flag_sign-conversion - Success
-- Using -Wsign-conversion: YES
-- Performing Test flag_shadow_local
-- Performing Test flag_shadow_local - Failed
-- Using -Wshadow=local: NO
-- Performing Test flag_old-style-cast
-- Performing Test flag_old-style-cast - Success
-- Using -Wold-style-cast: YES
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE) 
-- 
-- *** Crypto Summary ***
--   GNU TLS crypto enabled: OFF
--   OpenSSL crypto enabled: OFF
--   Native crypto enabled: ON
--   Default crypto: native
-- 
-- Looking for fseeko
-- Looking for fseeko - found
-- Looking for fseeko64
-- Looking for fseeko64 - not found
-- Looking for localtime_r
-- Looking for localtime_r - found
-- Looking for random
-- Looking for random - found
-- Performing Test HAVE_EXTERN_LONG_TIMEZONE
-- Performing Test HAVE_EXTERN_LONG_TIMEZONE - Success
-- Performing Test HAVE_EXTERN_TM_GMTOFF
-- Performing Test HAVE_EXTERN_TM_GMTOFF - Success
-- Performing Test LFS_WITHOUT_MACROS
-- Performing Test LFS_WITHOUT_MACROS - Failed
-- Performing Test LFS_WITH_MACROS
-- Performing Test LFS_WITH_MACROS - Success
-- Performing Test HAVE_MALLOC_INFO
-- Performing Test HAVE_MALLOC_INFO - Failed
-- Performing Test HAVE_OPEN_MEMSTREAM
-- Performing Test HAVE_OPEN_MEMSTREAM - Failed
-- Performing Test HAVE_LD_SCRIPT
-- Performing Test HAVE_LD_SCRIPT - Success
-- 
-- *** Summary ***
--   qpdf version: 11.4.0
--   build type: MinSizeRel
--   build shared libraries: ON
--   build static libraries: ON
--   build manual: OFF
--   compiler warnings are errors: OFF
--   QTC test coverage: OFF
--   system: Android
-- 
-- *** Options Summary ***
--   COMPILE_OPTIONS: -Wall;-Wconversion;-Wsign-conversion;$<$<COMPILE_LANGUAGE:CXX>:-Wold-style-cast>
--   COMPILE_DEFINITIONS: libqpdf_EXPORTS
--   INCLUDE_DIRECTORIES: /usr/include;$<BUILD_INTERFACE:/home/kartz/qpdf/include>;$<INSTALL_INTERFACE:include>;/home/kartz/qpdf/libqpdf;/home/kartz/qpdf/build/libqpdf
--   INTERFACE_INCLUDE_DIRECTORIES: $<BUILD_INTERFACE:/home/kartz/qpdf/include>;$<INSTALL_INTERFACE:include>
--   LINK_OPTIONS: -Wl,--version-script=/home/kartz/qpdf/build/libqpdf/libqpdf.map
--   LINK_LIBRARIES: /home/kartz/armv7a/build/libz.so;/home/kartz/armv7a/build/libjpeg.so
-- 
-- See above for crypto summary.
-- Configuring done (3.6s)
-- Generating done (0.2s)
-- Build files have been written to: /home/kartz/qpdf/build
```

Hi @jberkenbilt

I tried qpdf 12.1.0 today, and it seems that the misdetection is still there:

```
...
-- Performing Test LFS_WITHOUT_MACROS
-- Performing Test LFS_WITHOUT_MACROS - Failed
-- Performing Test LFS_WITH_MACROS
-- Performing Test LFS_WITH_MACROS - Success
-- Looking for fseeko
-- Looking for fseeko - found
-- Looking for fseeko64
-- Looking for fseeko64 - not found
...
```

I took a look at the generated `CMakeFiles/CMakeConfigureLog.yaml` file, which contains all the logs for the CMake configuration. Here is the specific part where qpdf tests for the existence of fseeko:

```
  -
    buildResult:
      variable: "HAVE_FSEEKO"
      cached: true
      stdout: |
        Change Dir: '/home/kartz/build/CMakeFiles/CMakeScratch/TryCompile-9BWJyJ'
        
        Run Build Command(s): /home/kartz/cmake-4.0.0-linux-x86_64/bin/cmake -E env VERBOSE=1 /opt/rh/devtoolset-9/root/usr/bin/gmake -f Makefile cmTC_2d258/fast
        /opt/rh/devtoolset-9/root/usr/bin/gmake  -f CMakeFiles/cmTC_2d258.dir/build.make CMakeFiles/cmTC_2d258.dir/build
        gmake[1]: Entering directory '/home/kartz/build/CMakeFiles/CMakeScratch/TryCompile-9BWJyJ'
        Building C object CMakeFiles/cmTC_2d258.dir/CheckSymbolExists.c.o
        /home/kartz/android-ndk-r27c/toolchains/llvm/prebuilt/linux-x86_64/bin/clang --target=armv7-none-linux-androideabi21 --sysroot=/home/kartz/android-ndk-r27c/toolchains/llvm/prebuilt/linux-x86_64/sysroot   -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -march=armv7-a -mthumb -Wformat   -fPIE -MD -MT CMakeFiles/cmTC_2d258.dir/CheckSymbolExists.c.o -MF CMakeFiles/cmTC_2d258.dir/CheckSymbolExists.c.o.d -o CMakeFiles/cmTC_2d258.dir/CheckSymbolExists.c.o -c /home/kartz/build/CMakeFiles/CMakeScratch/TryCompile-9BWJyJ/CheckSymbolExists.c
        Linking C executable cmTC_2d258
        /home/kartz/cmake-4.0.0-linux-x86_64/bin/cmake -E cmake_link_script CMakeFiles/cmTC_2d258.dir/link.txt --verbose=1
        /home/kartz/android-ndk-r27c/toolchains/llvm/prebuilt/linux-x86_64/bin/clang --target=armv7-none-linux-androideabi21 --sysroot=/home/kartz/android-ndk-r27c/toolchains/llvm/prebuilt/linux-x86_64/sysroot -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -march=armv7-a -mthumb -Wformat   -static-libstdc++ -Wl,--build-id=sha1 -Wl,--no-rosegment -Wl,--no-undefined-version -Wl,--fatal-warnings -Wl,--no-undefined -Qunused-arguments    CMakeFiles/cmTC_2d258.dir/CheckSymbolExists.c.o -o cmTC_2d258  -latomic -lm
        gmake[1]: Leaving directory '/home/kartz/build/CMakeFiles/CMakeScratch/TryCompile-9BWJyJ'
        
      exitCode: 0
  -
```

Looking at the compiler command invocation, I see that there is no `-D_FILE_OFFSET_BITS=64` anywhere in it:

```
/home/kartz/android-ndk-r27c/toolchains/llvm/prebuilt/linux-x86_64/bin/clang --target=armv7-none-linux-androideabi21 --sysroot=/home/kartz/android-ndk-r27c/toolchains/llvm/prebuilt/linux-x86_64/sysroot   -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -march=armv7-a -mthumb -Wformat   -fPIE -MD -MT 
```

I did reconfigure the project, but this time adding `-D_FILE_OFFSET_BITS=64` to `CMAKE_C_FLAGS`:

```bash
$ cmake -DCMAKE_C_FLAGS='-D_FILE_OFFSET_BITS=64' ...
-- Performing Test LFS_WITHOUT_MACROS
-- Performing Test LFS_WITHOUT_MACROS - Success
-- Performing Test LFS_WITH_MACROS
-- Performing Test LFS_WITH_MACROS - Success
-- Looking for fseeko
-- Looking for fseeko - not found
-- Looking for fseeko64
-- Looking for fseeko64 - not found
```

And now it was properly detected as not supported. Here is the output from `CMakeFiles/CMakeConfigureLog.yaml` for that specific part:

```
  -
    buildResult:
      variable: "HAVE_FSEEKO"
      cached: true
      stdout: |
        Change Dir: '/home/kartz/build/CMakeFiles/CMakeScratch/TryCompile-CsBIMb'
        
        Run Build Command(s): /home/kartz/cmake-4.0.0-linux-x86_64/bin/cmake -E env VERBOSE=1 /opt/rh/devtoolset-9/root/usr/bin/gmake -f Makefile cmTC_1e8ae/fast
        /opt/rh/devtoolset-9/root/usr/bin/gmake  -f CMakeFiles/cmTC_1e8ae.dir/build.make CMakeFiles/cmTC_1e8ae.dir/build
        gmake[1]: Entering directory '/home/kartz/build/CMakeFiles/CMakeScratch/TryCompile-CsBIMb'
        Building C object CMakeFiles/cmTC_1e8ae.dir/CheckSymbolExists.c.o
        /home/kartz/android-ndk-r27c/toolchains/llvm/prebuilt/linux-x86_64/bin/clang --target=armv7-none-linux-androideabi21 --sysroot=/home/kartz/android-ndk-r27c/toolchains/llvm/prebuilt/linux-x86_64/sysroot   -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -march=armv7-a -mthumb -Wformat -D_FILE_OFFSET_BITS=64  -fPIE -MD -MT CMakeFiles/cmTC_1e8ae.dir/CheckSymbolExists.c.o -MF CMakeFiles/cmTC_1e8ae.dir/CheckSymbolExists.c.o.d -o CMakeFiles/cmTC_1e8ae.dir/CheckSymbolExists.c.o -c /home/kartz/build/CMakeFiles/CMakeScratch/TryCompile-CsBIMb/CheckSymbolExists.c
        /home/kartz/build/CMakeFiles/CMakeScratch/TryCompile-CsBIMb/CheckSymbolExists.c:8:19: error: use of undeclared identifier 'fseeko'; did you mean 'fseek'?
            8 |   return ((int*)(&fseeko))[argc];
              |                   ^~~~~~
              |                   fseek
        /home/kartz/android-ndk-r27c/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/stdio.h:207:5: note: 'fseek' declared here
          207 | int fseek(FILE* _Nonnull __fp, long __offset, int __whence);
              |     ^
        1 error generated.
        gmake[1]: *** [CMakeFiles/cmTC_1e8ae.dir/build.make:82: CMakeFiles/cmTC_1e8ae.dir/CheckSymbolExists.c.o] Error 1
        gmake[1]: Leaving directory '/home/kartz/build/CMakeFiles/CMakeScratch/TryCompile-CsBIMb'
        gmake: *** [Makefile:134: cmTC_1e8ae/fast] Error 2
        
      exitCode: 2
  -
```

Based on this, it seems that on Android versions prior to 24, `fseeko()` has the same effect as `fseek()`. That's why the NDK toolchain doesn't bother omitting the function declaration when `-D_FILE_OFFSET_BITS=64` is not set because it assumes you already know this and that you don't care about it having no effect on the handling of large files.

When `-D_FILE_OFFSET_BITS=64` is set, you are explicitly asking for an `fseeko()` that supports large files, and since this is not supported on these old Android versions, the NDK toolchain then omits its declaration.

So the fix here seems to be always defining `_FILE_OFFSET_BITS` as `64` before checking for the existence of fseeko (at least, on armv7-a builds).

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -163,6 +163,7 @@ jobs:
           - test-alt-zlib
           - test-unsigned-char
           - test-c++-next
+          - build-android-ndk
     steps:
       - uses: actions/checkout@v4
       - name: ${{ matrix.script }}

--- a/build-scripts/build-android-ndk
+++ b/build-scripts/build-android-ndk
@@ -1,0 +1,41 @@
+#!/bin/bash
+set -xeo pipefail
+cd $(dirname $0)/..
+
+# Build a static libqpdf for android with an ABI version <= 24. This
+# is just to exercise that configuring for that platform works.
+
+declare -a android_flags
+android_flags=( \
+      -DCMAKE_TOOLCHAIN_FILE="$ANDROID_NDK/build/cmake/android.toolchain.cmake" \
+      -DANDROID_ABI=armeabi-v7a \
+      -DANDROID_PLATFORM=android-21 \
+)
+
+rm -rf android-work build-android
+mkdir android-work
+pushd android-work
+export WORK=$PWD
+git clone https://github.com/libjpeg-turbo/libjpeg-turbo
+cd libjpeg-turbo
+cmake -S . -B build \
+      "${android_flags[@]}" \
+      -DCMAKE_INSTALL_PREFIX="$WORK" \
+      -DENABLE_STATIC=ON \
+      -DENABLE_SHARED=OFF \
+      -DCMAKE_POSITION_INDEPENDENT_CODE=ON
+make -C build -j$(nproc)
+make -C build install
+popd
+
+export CMAKE_PREFIX_PATH=$WORK LDFLAGS="-L$WORK/lib"
+cmake -S . -B build-android \
+      "${android_flags[@]}" \
+      -DCMAKE_BUILD_TYPE=RelWithDebinfo \
+      -DUSE_IMPLICIT_CRYPTO=OFF \
+      -DBUILD_SHARED_LIBS=OFF \
+      -DBUILD_STATIC_LIBS=ON \
+      -DREQUIRE_CRYPTO_NATIVE=ON
+
+cmake --build build-android -t libqpdf --verbose -j$(nproc) -- -k
+file build-android/libqpdf/CMakeFiles/libqpdf_object.dir/QUtil.cc.o | grep 'ARM, EABI5'

--- a/libqpdf/CMakeLists.txt
+++ b/libqpdf/CMakeLists.txt
@@ -349,7 +349,11 @@ int main(int argc, char* argv[]) {
     LFS_WITHOUT_MACROS)
 
 check_c_source_compiles(
-"#define _FILE_OFFSET_BITS 64
+"#if !(defined(__ANDROID__) && __ANDROID_API__ <= 24)
+// Defining _FILE_OFFSET_BITS on Android <= 24 is known to
+// produce incorrect results.
+#define _FILE_OFFSET_BITS 64
+#endif
 #include <stdio.h>
 #include <sys/types.h>
 int main(int argc, char* argv[]) {
@@ -359,11 +363,6 @@ int main(int argc, char* argv[]) {
 if(LFS_WITH_MACROS AND NOT LFS_WITHOUT_MACROS)
   set(_FILE_OFFSET_BITS 64)
 endif()
-
-# fseeko may not exist with _FILE_OFFSET_BITS=64 but can exist with
-# _FILE_OFFSET_BITS unset or 32 (e.g. Android ARMv7 with NDK 26b and
-# API level < 24) so we need to test fseeko after testing for
-# _FILE_OFFSET_BITS. See https://github.com/qpdf/qpdf/issues/950.
 check_symbol_exists(fseeko "stdio.h" HAVE_FSEEKO)
 check_symbol_exists(fseeko64 "stdio.h" HAVE_FSEEKO64)
 

